// xslt-processor/dom/functions.js
"use strict";
// Copyright 2023 Design Liquido
// Copyright 2018 Johannes Wilm
// Copyright 2005 Google Inc.
// All Rights Reserved
Object.defineProperty(exports, "__esModule", { value: true });
exports.domCreateDTDSection = exports.domCreateDocumentFragment = exports.domCreateComment = exports.domCreateCDATASection = exports.domCreateElement = exports.domCreateTransformedTextNode = exports.domCreateTextNode = exports.domAppendTransformedChild = exports.domAppendChild = exports.domSetTransformedAttribute = exports.domSetAttribute = exports.domGetAttributeValue = void 0;
// Wrapper around DOM methods so we can condense their invocations.
function domGetAttributeValue(node, name) {
    return node.getAttributeValue(name);
}
exports.domGetAttributeValue = domGetAttributeValue;
function domSetAttribute(node, name, value) {
    return node.setAttribute(name, value);
}
exports.domSetAttribute = domSetAttribute;
function domSetTransformedAttribute(node, name, value) {
    return node.setTransformedAttribute(name, value);
}
exports.domSetTransformedAttribute = domSetTransformedAttribute;
function domAppendChild(node, child) {
    return node.appendChild(child);
}
exports.domAppendChild = domAppendChild;
function domAppendTransformedChild(node, child) {
    return node.appendTransformedChild(child);
}
exports.domAppendTransformedChild = domAppendTransformedChild;
function domCreateTextNode(node, text) {
    return node.createTextNode(text);
}
exports.domCreateTextNode = domCreateTextNode;
function domCreateTransformedTextNode(node, text) {
    return node.createTransformedTextNode(text);
}
exports.domCreateTransformedTextNode = domCreateTransformedTextNode;
function domCreateElement(doc, name) {
    return doc.createElement(name);
}
exports.domCreateElement = domCreateElement;
function domCreateCDATASection(doc, data) {
    return doc.createCDATASection(data);
}
exports.domCreateCDATASection = domCreateCDATASection;
function domCreateComment(doc, text) {
    return doc.createComment(text);
}
exports.domCreateComment = domCreateComment;
function domCreateDocumentFragment(doc) {
    return doc.createDocumentFragment();
}
exports.domCreateDocumentFragment = domCreateDocumentFragment;
function domCreateDTDSection(doc, data) {
    return doc.createDTDSection(data);
}
exports.domCreateDTDSection = domCreateDTDSection;
//XDocument.prototype = new XNode(DOM_DOCUMENT_NODE, '#document');
